client/src/components/geneExpression/geneSet.js

```diff
@@ -0,0 +1,187 @@
+import { memoize } from "../../util/dataframe/util";
+import Truncate from "../util/truncate";
+import * as globals from "../../globals";
+import GenesetMenus from "./menus/genesetMenus";
+
+@connect((state, ownProps) => {
+  return {
+    world: state.world,
+    userDefinedGenes: state.controls.userDefinedGenes,
+    userDefinedGenesLoading: state.controls.userDefinedGenesLoading,
+    isColorAccessor: state.colors.colorAccessor === ownProps.setName,
+  };
+})
+class GeneSet extends React.Component {
+  _memoGenesToUpper = memoize(this._genesToUpper, (arr) => arr);
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      isOpen: false,
+      toggleSummaryHisto: false,
+    };
+  }
+
+  _genesToUpper = (listGenes) => {
+    // Has to be a Map to preserve index
+    const upperGenes = new Map();
+    for (let i = 0, { length } = listGenes; i < length; i += 1) {
+      upperGenes.set(listGenes[i].toUpperCase(), i);
+    }
+
+    return upperGenes;
+  };
+
+  fetchGenes = () => {
+    const { world, dispatch, setGenes } = this.props;
+    const varIndexName = world.schema.annotations.var.index;
+
+    const worldGenes = world.varAnnotations.col(varIndexName).asArray();
+
+    const upperGenes = this._genesToUpper(setGenes);
+    const upperWorldGenes = this._memoGenesToUpper(worldGenes);
+
+    dispatch({ type: "bulk user defined gene start" });
+
+    Promise.all(
+      [...upperGenes.keys()].map((upperGene) => {
+        const indexOfGene = upperWorldGenes.get(upperGene);
+
+        if (indexOfGene === undefined) {
+          console.log("found a gene that doesn't appear to be a valid name");
+        }
+        return dispatch(
+          actions.requestUserDefinedGene(worldGenes[indexOfGene])
+        );
+      })
+    ).then(
+      () => dispatch({ type: "bulk user defined gene complete" }),
+      () => dispatch({ type: "bulk user defined gene error" })
+    );
+
+    return undefined;
+  };
+
+  onGenesetMenuClick = () => {
+    const { isOpen } = this.state;
+    this.setState({ isOpen: !isOpen });
+  };
+
+  onColorChangeClick = () => {
+    const { dispatch, setName } = this.props;
+    dispatch({
+      type: "color by gene set",
+      colorAccessor: setName,
+    });
+  };
+
+  toggleSummaryHisto = () => {
+    const { toggleSummaryHisto } = this.state;
+    this.setState({ toggleSummaryHisto: !toggleSummaryHisto });
+  };
+
+  render() {
+    const { setName, setGenes, isDiffexp } = this.props;
+    const { isOpen, toggleSummaryHisto } = this.state;
+    return (
+      <div style={{ marginBottom: 3 }}>
+        <div
+          style={{
+            display: "flex",
+            justifyContent: "space-between",
+            alignItems: "baseline",
+          }}
+        >
+          <span
+            role="menuitem"
+            tabIndex="0"
+            data-testclass="geneset-expand"
+            data-testid={`${setName}:geneset-expand`}
+            onKeyPress={/* todo_genesets */ () => {}}
```

Comment:
is this TODO captured in an issue, or should it be in this PR?

Reply:
https://github.com/chanzuckerberg/cellxgene/issues/2101

Reply:
Same as `gene.js`